tests/conftest.py

```diff
@@ -99,48 +126,50 @@ def set_docker_images():
 
 
 @pytest.fixture(scope='session')
-def session_id():
+def session_id(namespace):
+    if namespace:
+        return namespace
     return random_id(length=5)
 
 
 @pytest.fixture(scope='session')
-def set_namespace(request, session_id, load_config):
-
+def set_namespace(request, session_id, load_config, delete_ns):
     v1 = CoreV1ApiClient()
 
-    def _setup_namespace():
+    def setup_namespace():
 
         if testconfig['namespace'] == '':
             testconfig['namespace'] = session_id
 
         print("\nRun tests in namespace: {0}".format(testconfig['namespace']))
-        print("Kibana URL: http://kibana.spacemesh.io/app/kibana#/discover?_g=(refreshInterval:(pause:!t,value:0))&_a=("
-              "columns:!(M),filters:!(('$state':(store:appState),meta:(alias:!n,disabled:!f,index:'8a91d9d0-c24f-11e9-9"
-              "a59-a76b835079b3',key:kubernetes.namespace_name,negate:!f,params:(query:{0},type:phrase),type:phrase,val"
-              "ue:{0}),query:(match:(kubernetes.namespace_name:(query:{0},type:phrase))))))"
-              .format(testconfig['namespace']), "\n")
         namespaces_list = [ns.metadata.name for ns in v1.list_namespace().items]
         if testconfig['namespace'] in namespaces_list:
+            pytest.delete_namespace = False
             return
 
         body = client.V1Namespace()
         body.metadata = client.V1ObjectMeta(name=testconfig['namespace'])
         v1.create_namespace(body)
 
-    def fin():
+    def fin(conditions):
         # On teardown we wish to report on pods that were restarted by k8s during the test
         restarted_pods = pod.check_for_restarted_pods(testconfig['namespace'])
         if restarted_pods:
             print('\n\nAttention!!! The following pods were restarted during test: {0}\n\n'.format(restarted_pods))
-
-        if hasattr(request, 'param') and request.param == 'doNotDeleteNameSpace':
-            print("\nDo not delete namespace: {0}".format(testconfig['namespace']))
-        else:
+        if all(condition is True for condition in conditions):
             print("\nDeleting test namespace: {0}".format(testconfig['namespace']))
             v1.delete_namespace(name=testconfig['namespace'], body=client.V1DeleteOptions())
+        else:
+            exit_msg = f"NOTICE!!! namespace {session_id} was not deleted!!!\n" \
+                       f"please make sure to delete it after use:\n" \
+                       f"kubectl delete namespace {session_id}"
+            print(exit_msg)
 
-    request.addfinalizer(fin)
-    return _setup_namespace()
+    setup_namespace()
+    yield
+    # delete namespace if was not mentioned otherwise in the command line args and if global arg is set to
+    # True (pytest.delete_namespace)
+    fin([delete_ns, pytest.delete_namespace])
```

Comment:
This code is organized in a VERY confusing way. Defining functions and then calling them immediately is bad form.

We should put the code directly in `set_namespace()` - first the code that's currently in `setup_namespace()` (replacing the `return` there with an `else`), then the `yield` keyword, along with a comment like 
```python
# pytest considers any code after `yield` to be teardown code
# see: https://docs.pytest.org/en/reorganize-docs/yieldfixture.html
```

and then the code that's currently in `fin()`.

Reply:
Alternatively, define the two methods outside of `set_namespace()` and make it just call `setup_namespace()`, then `yield` + the comment, then `fin()`.

That's actually my preferred solution.

Reply:
About the code - this is Itzik code, I just added the yield command, my plan was to rewrite this function so I didn't bother unpacking the nested functions.

I've used this yield keyword in other fixtures should I add this comment there as well?
I think that it's obvious for someone who knows what yield means and python in general.

Reply:
fixed